torchquad/integration/newton_cotes.py

```diff
@@ -47,8 +48,16 @@ def calculate_result(self, function_values, dim, n_per_dim, hs):
             backend tensor: Quadrature result
         """
         # Reshape the output to be [N,N,...] points instead of [dim*N] points
-        function_values = function_values.reshape([n_per_dim] * dim)
-
+        self.integrand_shape = function_values.shape[1:]
+        if len(self.integrand_shape) == 0 or (len(self.integrand_shape) == 1 and self.integrand_shape[0] == 1):
```

Comment:
In general, I was unsure what the best way to handle checking for "multi-dimensionality" of the integrand (if that is even an appropriate way to refer to this) so definitely open to suggestions here

Reply:
Good question. Previously I had thought we'd move this into type of integrator into different file but we can also put it into the existing if that works ok.

There are two main concerns here for me:
1) Don't break backwards compatibility, we don't want to break existing code using torchquad ideally.
2) Make it intuitive, elegant and consistent with current API.

So, old code like, e.g.

```py
def f(x):
    return torch.exp(x) * torch.pow(x, 2)
integration_domain = [[0, 2]]
tp = Trapezoid()
result = tp.integrate(f, dim=1, N=101, integration_domain=integration_domain)
```

Should still work as expected. Following what you have started to implement, maybe we can aim for something like:

Let's say my integrand is

```py
def f(x,my_other_parameter):
    return torch.exp(x) * torch.pow(x, 2) * my_other_parameter
```
with `my_other_parameter` in [1,10]

Then, the goal would be to allow me to evaluate it in one call, right?
So, I would expect something like:

```py
integration_domain = [[0, 2]]
# Define a function that maps index of evaluation to additional parameter
parameter = lambda idx: idx + 1
# Define how often to evaluate
N_parameter = 10
# Define integrator as always
tp = Trapezoid()

# Now I pass them, before the would have defaulted to None & 1 
# And internally now this triggers an additional first dimension of the `function_value` thing we discussed above
# with size 10 etc.
result = tp.integrate(f, dim=1, N=101, integration_domain=integration_domain, additional_f_arg=parameter,N_additional_f_evals=N_parameter)
```

Get the idea? What do you think?

Reply:
(there may be better names for these things, just trying to illustrate)

Reply:
Note that in your case, you could e.g. have some constants or whatever in an array and define your parameter functions as
```py
parameter = lambda idx: your_precomputed_constants[idx]
```

Reply:
No concrete objection, I was just trying to explore how the user perspective on this might look and how do deal with the dimensionality. How would you currently use it, is it different from my example?

Reply:
I did add new tests that should display what I'm going for.  I use `einsum` to do what I need to.  i.e you have an integrand of size `KxL` and `N` evaluation points from the trapezoidal rule - then you need to broadcast the `N` over the `KxL` elementwise, and I landed on einstein summation to do that: https://github.com/esa/torchquad/pull/160/files#diff-fe1b22e294ae47b16639a2139938eeec68d387dfa25d702da595cb2e75ef1bafR107-R109

Reply:
But maybe you would have a revursive thing going where teh entries in the output integrand matrix depend on one another, and then einstein summation would not work, I think.  So if the top left entry were some constant and then the next entry going right were a function of that.  Just guess/imagining a different use-case than mine, which is (basically) scalar multiplication of exponentials.